File: sharedComposables/src/commonMain/kotlin/dev/mateusz1913/f1results/composable/circuit/CircuitScreen.kt
package dev.mateusz1913.f1results.composable.circuit

import androidx.compose.foundation.layout.*
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.mateusz1913.f1results.composable.common.InfoRow
import dev.mateusz1913.f1results.composable.common.Loading
import dev.mateusz1913.f1results.composable.di.getViewModelInstance
import dev.mateusz1913.f1results.viewmodel.CircuitViewModel
import org.koin.core.parameter.parametersOf

@Composable
fun CircuitScreen(
    circuitId: String,
    circuitViewModel: CircuitViewModel = getViewModelInstance { parametersOf(circuitId) }
) {
    val circuitState = circuitViewModel.circuitState.collectAsState()
    val circuit = circuitState.value.circuit
    if (circuitState.value.isFetching) {
        Loading()
        return
    }
    if (circuit == null) {
        Text("No circuit")
        return
    }
    CircuitLayoutWithMap(circuit) {
        Box(
            modifier = Modifier.height(128.dp)
        ) {
            Column(
                modifier = Modifier.fillMaxSize(),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center,
            ) {
                Box(modifier = Modifier.padding(top = 24.dp, start = 20.dp, end = 20.dp)) {
                    InfoRow(
                        label = "Circuit: ",
                        value = circuit.circuitName,
                        fontSize = 20.sp,
                        fontWeight = FontWeight.Bold
                    )
                }
                Row {
                    Column(
                        modifier = Modifier.weight(0.5f).padding(vertical = 10.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        InfoRow(
                            label = "Locality: ",
                            value = circuit.location.locality,
                            fontSize = 16.sp,
                            fontWeight = FontWeight.SemiBold
                        )
                    }
                    Column(
                        modifier = Modifier.weight(0.5f).padding(vertical = 10.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        InfoRow(
                            label = "Country: ",
                            value = circuit.location.country,
                            fontSize = 16.sp,
                            fontWeight = FontWeight.SemiBold
                        )
                    }
                }
            }
        }
    }
}

File: swiftApp/Shared/Presentation/Circuit/CircuitScreen.swift
import CoreLocation
import SwiftUI
import shared

struct CircuitScreen: View {
    @ObservedObject var circuitState: CircuitState
    
    var body: some View {
        if circuitState.circuitIsFetching && circuitState.circuit == nil {
            loadingBody
        } else if circuitState.circuit == nil {
            emptyBody
        } else {
            resultBody
        }
    }
    
    private var emptyBody: some View {
        VStack {
            Text("No circuit")
        }
        .fillMaxSize()
    }
    
    private var loadingBody: some View {
        VStack {
            Loading()
        }
        .fillMaxSize()
    }
    
    private var resultBody: some View {
        circuitState.circuit.map { circuit in
            VStack {
                if let latString = circuit.location.lat, let longString = circuit.location.long_, let latitude = Double(latString), let longitude = Double(longString) {
                    CircuitLayoutWithMap(latitude: latitude, longitude: longitude) {
                        VStack {
                            VStack {
                                InfoRow(fontSize: 20, fontWeight: .bold, label: "Circuit: ", value: circuit.circuitName)
                            }
                            .padding(.init(top: 24, leading: 20, bottom: 0, trailing: 20))
                            HStack {
                                VStack {
                                    InfoRow(fontSize: 16, fontWeight: .semibold, label: "Locality: ", value: circuit.location.locality)
                                }
                                .fillMaxWidth()
                                .padding(.vertical, 10)
                                VStack {
                                    InfoRow(fontSize: 16, fontWeight: .semibold, label: "Country: ", value: circuit.location.country)
                                }
                                .fillMaxWidth()
                                .padding(.vertical, 10)
                            }
                        }
                        .fillMaxWidth()
                    }
                }
            }
            .fillMaxSize()
        }
    }
}
